File: stone_lib/src/main/java/com/github/klee0kai/stone/types/wrappers/ProviderWrapper.java
package com.github.klee0kai.stone.types.wrappers;

import javax.inject.Provider;

public interface ProviderWrapper {

    /**
     * Provide wrapped object.
     *
     * @param wrapperCl        type of wrapper
     * @param originalProvider original object provider
     * @param <Wr>             type of wrapper
     * @param <T>              type of providing original object
     * @return wrapped object provider
     */
    <Wr, T> Wr wrap(Class<Wr> wrapperCl, Provider<T> originalProvider);

}

File: stone_multiplatform/src/commonMain/kotlin/stone/wrappers/creators/ProviderWrapper.kt
package stone.wrappers.creators

import javax.inject.Provider

interface ProviderWrapper {
    /**
     * Provide wrapped object.
     *
     * @param wrapperCl        type of wrapper
     * @param originalProvider original object provider
     * @param <Wr>             type of wrapper
     * @param <T>              type of providing original object
     * @return wrapped object provider
    </T></Wr> */
    fun <Wr, T> wrap(wrapperCl: Class<Wr?>?, originalProvider: Provider<T?>?): Wr?
}
